File: pipeline/shared/src/commonMain/kotlin/com/informedcitizen/pipeline/fetch/FetchVotes.kt
package com.informedcitizen.pipeline.fetch

import com.informedcitizen.pipeline.ErrorCollector
import com.informedcitizen.pipeline.http.SenateVotesApiException
import com.informedcitizen.pipeline.http.SenateVotesClient
import com.informedcitizen.pipeline.model.Chamber
import com.informedcitizen.pipeline.model.RollCallVote
import com.informedcitizen.pipeline.model.VoteRef
import com.informedcitizen.pipeline.model.VotesIndex
import kotlinx.coroutines.CancellationException

/**
 * Fetch roll-call votes (both chambers) and publish per-vote JSON plus
 * a per-Congress index. Direct port of Python `fetch_votes.py`.
 *
 * Both sources are keyless: Senate votes come from senate.gov LIS XML
 * (menu + detail documents), House votes from clerk.house.gov EVS XML.
 * The House publishes no machine-readable vote menu, so new House
 * rolls are discovered by probing sequentially past the rolls already
 * on disk until the first 404 (roll numbers are dense per
 * calendar-year session).
 *
 * Runs are incremental and self-resuming: a roll call already on disk
 * is never refetched (published roll calls are immutable), so each run
 * costs the Senate vote menu(s), the two legislators YAMLs, one detail
 * XML per *new* vote, and a handful of House probe fetches. The index
 * is rebuilt from the files on disk at the end of every run, so a
 * crashed run heals on the next one.
 */

const val LEGISLATORS_CURRENT_YAML_URL: String =
    "https://raw.githubusercontent.com/" +
        "unitedstates/congress-legislators/main/legislators-current.yaml"

const val LEGISLATORS_HISTORICAL_YAML_URL: String =
    "https://raw.githubusercontent.com/" +
        "unitedstates/congress-legislators/main/legislators-historical.yaml"

// Both sessions are always attempted; the menu for a session that
// hasn't started yet 404s and is skipped (same pattern as the Senate
// schedule walk in buildSessionCalendar).
private val SENATE_SESSIONS = listOf(1, 2)

// House sessions are probed the same way: a session that hasn't
// started 404s on its first roll and costs a single fetch.
private val HOUSE_SESSIONS = listOf(1, 2)

// A clerk.house.gov outage or format drift looks like a run of
// consecutive parse failures; a single bad roll has parseable
// successors that reset the counter. The cap keeps a
// 200-with-error-page response from probing forever.
internal const val MAX_CONSECUTIVE_HOUSE_ERRORS: Int = 3

/** Mirrors Python `fetch_votes.py`'s `--max-new` default. */
const val FETCH_VOTES_MAX_NEW_DEFAULT: Int = 1000

/** Unrecoverable fetch failure — the CLI reports it and exits 1. */
class FetchVotesException(
    message: String,
    cause: Throwable? = null,
) : RuntimeException(message, cause)

data class FetchVotesResult(
    val congress: Int,
    /** Sessions whose Senate vote menu was published and parsed. */
    val sessions: List<Int>,
    /** Roll calls listed across all fetched Senate menus. */
    val totalListed: Int,
    /** Total new votes saved this run (both chambers). */
    val fetched: Int,
    val senateFetched: Int,
    val houseFetched: Int,
    /** Total votes already on disk (both chambers). */
    val skipped: Int,
    /** House candidate-ballot votes (Speaker elections) skipped. */
    val unsupported: Int,
    val index: VotesIndex,
    /** True iff the bills manifest's vote refs changed and it was rewritten. */
    val billManifestRefreshed: Boolean,
)

/**
 * Streaming progress hooks. Defaulted to no-ops so the orchestrator
 * stays testable without log noise; the CLI wraps each with a `println`.
 */
data class FetchVotesProgress(
    val onMenus: (totalListed: Int, sessions: List<Int>) -> Unit = { _, _ -> },
    val onVoteSaved: (vote: RollCallVote) -> Unit = {},
    val onMaxNewReached: (maxNew: Int) -> Unit = {},
    /** The House probe is starting (mirrors Python's banner line). */
    val onHouseProbeStart: () -> Unit = {},
    /** The Senate fetch exhausted the `--max-new` budget; House deferred. */
    val onHouseDeferred: () -> Unit = {},
    /** A candidate-ballot vote was skipped, e.g. `voteKey = "119-1-2"`. */
    val onUnsupportedVote: (voteKey: String, message: String) -> Unit = { _, _ -> },
    /** [MAX_CONSECUTIVE_HOUSE_ERRORS] hit; the session probe stopped. */
    val onConsecutiveFailures: (count: Int, session: Int) -> Unit = { _, _ -> },
)

/**
 * Union of the current and historical legislators YAMLs. Mirrors
 * Python `fetch_votes.build_lis_to_bioguide`.
 *
 * Historical is required, not a nicety: a senator who leaves
 * mid-Congress moves to legislators-historical.yaml while their roll
 * calls stay published. Current entries win on the (unobserved) chance
 * of a conflict.
 */
suspend fun buildLisToBioguide(
    client: SenateVotesClient,
    currentYamlUrl: String = LEGISLATORS_CURRENT_YAML_URL,
    historicalYamlUrl: String = LEGISLATORS_HISTORICAL_YAML_URL,
): Map<String, String> {
    val historical = parseLisToBioguideYaml(client.fetch(historicalYamlUrl))
    val current = parseLisToBioguideYaml(client.fetch(currentYamlUrl))
    return historical + current
}

/**
 * Fetch and parse the vote menu for each published session. Mirrors
 * Python `fetch_votes.fetch_senate_menus`.
 *
 * A 404 means the session hasn't started (or its menu isn't up yet)
 * and is skipped; any other HTTP failure propagates. Throws
 * [FetchVotesException] if no session menu is available at all — that
 * means the congress number is wrong or senate.gov changed its layout,
 * either way nothing useful can be published.
 */
suspend fun fetchSenateMenus(client: SenateVotesClient, congress: Int): List<SenateVoteMenu> {
    val menus = mutableListOf<SenateVoteMenu>()
    for (session in SENATE_SESSIONS) {
        val url = senateVoteMenuUrl(congress, session)
        val text = try {
            client.fetch(url)
        } catch (e: SenateVotesApiException) {
            if (e.status == 404) continue
            throw e
        }
        val menu = parseSenateVoteMenu(text)
        if (menu.congress != congress || menu.session != session) {
            throw FetchVotesException(
                "vote menu at $url identifies itself as congress " +
                    "${menu.congress} session ${menu.session}",
            )
        }
        menus += menu
    }
    if (menus.isEmpty()) {
        throw FetchVotesException("no Senate vote menu available for Congress $congress")
    }
    return menus
}

private data class HouseFetchResult(
    val fetched: Int,
    val skipped: Int,
    val unsupported: Int,
)

/**
 * Probe clerk.house.gov for new House roll calls and save them.
 * Mirrors Python `fetch_votes.fetch_new_house_votes`.
 *
 * The House publishes no vote menu, but roll numbers are dense and
 * sequential per session, so each session is walked from roll 1: rolls
 * already on disk are skipped for free, missing ones are fetched, and
 * the first 404 is the end of the published roll calls. Non-404 fetch
 * failures are recorded and stop the session probe (an outage is
 * indistinguishable from the end of the rolls; the next run re-walks
 * from roll 1 with free existence checks). Candidate-ballot votes
 * (Speaker elections) cannot be expressed as yea/nay positions; they
 * are skipped — and re-probed every run, a couple of cheap fetches per
 * Congress — rather than recorded, keeping file existence as the only
 * resume cursor.
 */
private suspend fun fetchNewHouseVotes(
    client: SenateVotesClient,
    store: FileVotesStore,
    congress: Int,
    errors: ErrorCollector,
    maxNew: Int,
    progress: FetchVotesProgress,
): HouseFetchResult {
    var fetched = 0
    var skipped = 0
    var unsupported = 0
    for (session in HOUSE_SESSIONS) {
        var consecutiveErrors = 0
        var rollNumber = 0
        session@ while (true) {
            rollNumber += 1
            if (store.voteExists(congress, Chamber.HOUSE, session, rollNumber)) {
                skipped++
                continue
            }
            if (fetched >= maxNew) {
                progress.onMaxNewReached(maxNew)
                return HouseFetchResult(fetched, skipped, unsupported)
            }
            val url = houseVoteSourceUrl(congress, session, rollNumber)
            val voteKey = "$congress-$session-$rollNumber"
            val text = try {
                client.fetch(url)
            } catch (e: CancellationException) {
                throw e
            } catch (e: Throwable) {
                if (e is SenateVotesApiException && e.status == 404) {
                    break@session // end of the published roll calls for this session
                }
                // Outage vs. end is indistinguishable; retry next run.
                errors.record(
                    kind = "house_vote",
                    identifier = voteKey,
                    errorClass = e::class.simpleName ?: "Throwable",
                    message = e.message ?: e.toString(),
                    url = url,
                )
                break@session
            }
            val vote = try {
                parseHouseVote(text)
            } catch (e: UnsupportedVoteException) {
                unsupported++
                consecutiveErrors = 0
                progress.onUnsupportedVote(voteKey, e.message.orEmpty())
                continue
            } catch (e: CancellationException) {
                throw e
            } catch (e: Throwable) {
                // Per-vote isolation: record and retry next run.
                errors.record(
                    kind = "house_vote",
                    identifier = voteKey,
                    errorClass = e::class.simpleName ?: "Throwable",
                    message = e.message ?: e.toString(),
                    url = url,
                )
                consecutiveErrors++
                if (consecutiveErrors >= MAX_CONSECUTIVE_HOUSE_ERRORS) {
                    progress.onConsecutiveFailures(consecutiveErrors, session)
                    break@session
                }
                continue
            }
            if (vote.congress != congress || vote.session != session || vote.rollNumber != rollNumber) {
                errors.record(
                    kind = "house_vote",
                    identifier = voteKey,
                    errorClass = "IllegalArgumentException",
                    message = "detail XML identifies itself as ${vote.id}",
                    url = url,
                )
                break@session // the year->session mapping is off; nothing later can be right
            }
            consecutiveErrors = 0
            store.saveVote(vote)
            fetched++
            progress.onVoteSaved(vote)
        }
    }
    return HouseFetchResult(fetched, skipped, unsupported)
}

/**
 * Re-attach vote refs to the bills manifest after new votes land.
 * Mirrors Python `fetch_votes.refresh_bill_vote_refs`.
 *
 * Bills leave fetch-bills' refresh window ~60 days after their last
 * action, so this pass — not the bills workflow — is what links votes
 * that arrive later (or between bill runs) to their bills. Returns true
 * when the manifest changed and was rewritten; no-ops when there is no
 * manifest to enrich (votes can backfill before bills on a fresh tree).
 */
fun refreshBillVoteRefs(
    manifestStore: FileBillsManifestStore,
    congress: Int,
    refs: List<VoteRef>,
    nowIso: String,
): Boolean {
    val bills = manifestStore.load(congress)?.bills ?: return false
    val enriched = attachVoteRefs(bills, refs)
    if (enriched == bills) return false
    manifestStore.save(congress, enriched, nowIso)
    return true
}

/**
 * Full fetch-votes orchestrator. Mirrors Python `fetch_votes.main`:
 * fetch the Senate session menus and the lis→bioguide map (both fatal
 * on failure), fetch/parse/save every menu vote not already on disk (a
 * vote that fails is recorded in [errors] and left off disk, so the
 * next run retries it; detail XML whose self-identified session/roll
 * disagrees with the menu is rejected rather than published under the
 * wrong id), probe clerk.house.gov for new House rolls with the
 * remaining [maxNew] budget, then rebuild the index from every vote
 * file on disk and — when a [manifestStore] is supplied — refresh the
 * bills manifest's derived vote refs ([refreshBillVoteRefs]).
 *
 * [maxNew] caps new fetches per run, shared across both chambers
 * (Senate first, remainder to the House); a capped run is harmless —
 * the next run picks up where it left off.
 */
suspend fun fetchVotes(
    client: SenateVotesClient,
    store: FileVotesStore,
    congress: Int,
    nowIso: String,
    errors: ErrorCollector,
    manifestStore: FileBillsManifestStore? = null,
    maxNew: Int = FETCH_VOTES_MAX_NEW_DEFAULT,
    progress: FetchVotesProgress = FetchVotesProgress(),
    currentYamlUrl: String = LEGISLATORS_CURRENT_YAML_URL,
    historicalYamlUrl: String = LEGISLATORS_HISTORICAL_YAML_URL,
): FetchVotesResult {
    val menus = fetchSenateMenus(client, congress)
    val lisToBioguide = buildLisToBioguide(client, currentYamlUrl, historicalYamlUrl)
    val totalListed = menus.sumOf { it.voteNumbers.size }
    progress.onMenus(totalListed, menus.map { it.session })

    var senateFetched = 0
    var senateSkipped = 0
    outer@ for (menu in menus) {
        for (rollNumber in menu.voteNumbers) {
            if (store.voteExists(congress, Chamber.SENATE, menu.session, rollNumber)) {
                senateSkipped++
                continue
            }
            if (senateFetched >= maxNew) {
                progress.onMaxNewReached(maxNew)
                break@outer
            }
            val url = senateVoteSourceUrl(congress, menu.session, rollNumber)
            val vote = try {
                val parsed = parseSenateVote(client.fetch(url), lisToBioguide)
                check(parsed.session == menu.session && parsed.rollNumber == rollNumber) {
                    "detail XML identifies itself as session ${parsed.session} " +
                        "roll ${parsed.rollNumber}"
                }
                parsed
            } catch (e: CancellationException) {
                throw e
            } catch (e: Throwable) {
                // Per-vote isolation: record and retry next run.
                errors.record(
                    kind = "senate_vote",
                    identifier = "$congress-${menu.session}-$rollNumber",
                    errorClass = e::class.simpleName ?: "Throwable",
                    message = e.message ?: e.toString(),
                    url = url,
                )
                continue
            }
            store.saveVote(vote)
            senateFetched++
            progress.onVoteSaved(vote)
        }
    }

    val houseBudget = maxNew - senateFetched
    val house = if (houseBudget > 0) {
        progress.onHouseProbeStart()
        fetchNewHouseVotes(client, store, congress, errors, houseBudget, progress)
    } else {
        progress.onHouseDeferred()
        HouseFetchResult(fetched = 0, skipped = 0, unsupported = 0)
    }

    val refs = store.loadVotes(congress).map(::buildVoteRef)
    val index = buildVotesIndex(congress, refs, nowIso)
    store.saveIndex(index)
    val billManifestRefreshed = manifestStore != null &&
        refreshBillVoteRefs(manifestStore, congress, index.votes, nowIso)
    return FetchVotesResult(
        congress = congress,
        sessions = menus.map { it.session },
        totalListed = totalListed,
        fetched = senateFetched + house.fetched,
        senateFetched = senateFetched,
        houseFetched = house.fetched,
        skipped = senateSkipped + house.skipped,
        unsupported = house.unsupported,
        index = index,
        billManifestRefreshed = billManifestRefreshed,
    )
}

File: data-pipeline/scripts/fetch_votes.py
"""Fetch roll-call votes (both chambers) and publish per-vote JSON plus a per-Congress index.

Both sources are keyless: Senate votes come from senate.gov LIS XML (menu +
detail documents), House votes from clerk.house.gov EVS XML. The House
publishes no machine-readable vote menu, so new House rolls are discovered by
probing sequentially past the rolls already on disk until the first 404 (roll
numbers are dense per calendar-year session).

Layout under docs/data/ (matches the RollCallVote/VotesIndex wire models in
pipeline:shared):

    votes/congress119/senate-1-618.json   one file per roll call, with positions
    votes/congress119/house-1-17.json
    congress119_votes.json                index: every vote minus positions

Runs are incremental and self-resuming: a roll call already on disk is never
refetched (published roll calls are immutable), so each run costs the Senate
vote menu(s), the two legislators YAMLs, one detail XML per *new* vote, and a
handful of House probe fetches. The index is rebuilt from the files on disk at
the end of every run, so a crashed run heals on the next one.

Run locally (no key required):
    python data-pipeline/scripts/fetch_votes.py
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path
from typing import Any

import requests

import _common
from _common import ErrorCollector, LEGISLATORS_CURRENT_YAML_URL, current_congress, now_iso
from _votes import (
    CHAMBER_HOUSE,
    CHAMBER_SENATE,
    UnsupportedVoteError,
    build_vote_ref,
    build_votes_index,
    house_vote_source_url,
    parse_house_vote,
    parse_lis_to_bioguide_yaml,
    parse_senate_vote,
    parse_senate_vote_menu,
    senate_vote_menu_url,
    senate_vote_source_url,
    vote_file_relpath,
)

LEGISLATORS_HISTORICAL_YAML_URL = (
    "https://raw.githubusercontent.com/"
    "unitedstates/congress-legislators/main/legislators-historical.yaml"
)

# Both sessions are always attempted; the menu for a session that hasn't
# started yet 404s and is skipped (same pattern as the Senate schedule walk
# in build_session_calendar.py).
SENATE_SESSIONS = (1, 2)

# House sessions are probed the same way: a session that hasn't started 404s
# on its first roll and costs a single fetch.
HOUSE_SESSIONS = (1, 2)

# A clerk.house.gov outage or format drift looks like a run of consecutive
# parse failures; a single bad roll has parseable successors that reset the
# counter. The cap keeps a 200-with-error-page response from probing forever.
MAX_CONSECUTIVE_HOUSE_ERRORS = 3

TIMEOUT_SECONDS = 30


def _fetch(url: str) -> str:
    response = requests.get(
        url,
        headers={"User-Agent": _common.USER_AGENT},
        timeout=TIMEOUT_SECONDS,
    )
    response.raise_for_status()
    response.encoding = "utf-8"
    return response.text


# ---------- IO (paths resolve against _common.OUTPUT_DIR at call time) -----


def votes_dir(congress: int) -> Path:
    return _common.OUTPUT_DIR / "votes" / f"congress{congress}"


def vote_path(congress: int, chamber: str, session: int, roll_number: int) -> Path:
    return _common.OUTPUT_DIR / vote_file_relpath(congress, chamber, session, roll_number)


def votes_index_path(congress: int) -> Path:
    return _common.OUTPUT_DIR / f"congress{congress}_votes.json"


def rebuild_votes_index(congress: int) -> dict[str, Any]:
    """Rebuild congress{N}_votes.json from every vote file on disk."""
    refs = []
    directory = votes_dir(congress)
    if directory.exists():
        for path in sorted(directory.glob("*.json")):
            with path.open("r", encoding="utf-8") as f:
                refs.append(build_vote_ref(json.load(f)))
    payload = build_votes_index(congress, refs, now_iso())
    _common._write_json(votes_index_path(congress), payload)
    return payload


# ---------- fetching --------------------------------------------------------


def build_lis_to_bioguide() -> dict[str, str]:
    """Union of the current and historical legislators YAMLs.

    Historical is required, not a nicety: a senator who leaves mid-Congress
    moves to legislators-historical.yaml while their roll calls stay
    published. Current entries win on the (unobserved) chance of a conflict.
    """
    historical = parse_lis_to_bioguide_yaml(_fetch(LEGISLATORS_HISTORICAL_YAML_URL))
    current = parse_lis_to_bioguide_yaml(_fetch(LEGISLATORS_CURRENT_YAML_URL))
    return {**historical, **current}


def fetch_senate_menus(congress: int) -> list[dict[str, Any]]:
    """Fetch and parse the vote menu for each published session.

    A 404 means the session hasn't started (or its menu isn't up yet) and is
    skipped; any other HTTP failure propagates. Raises if no session menu is
    available at all — that means the congress number is wrong or senate.gov
    changed its layout, either way nothing useful can be published.
    """
    menus: list[dict[str, Any]] = []
    for session in SENATE_SESSIONS:
        url = senate_vote_menu_url(congress, session)
        try:
            text = _fetch(url)
        except requests.HTTPError as exc:
            if exc.response is not None and exc.response.status_code == 404:
                continue
            raise
        menu = parse_senate_vote_menu(text)
        if menu["congress"] != congress or menu["session"] != session:
            raise RuntimeError(
                f"vote menu at {url} identifies itself as congress "
                f"{menu['congress']} session {menu['session']}"
            )
        menus.append(menu)
    if not menus:
        raise RuntimeError(f"no Senate vote menu available for Congress {congress}")
    return menus


def fetch_new_senate_votes(
    congress: int,
    menus: list[dict[str, Any]],
    lis_to_bioguide: dict[str, str],
    errors: ErrorCollector,
    max_new: int,
) -> tuple[int, int]:
    """Fetch, parse, and save every menu vote not already on disk.

    Returns (fetched, skipped). A vote that fails to fetch or parse is
    recorded in ``errors`` and left off disk, so the next run retries it.
    """
    fetched = 0
    skipped = 0
    for menu in menus:
        session = menu["session"]
        for roll_number in menu["vote_numbers"]:
            path = vote_path(congress, CHAMBER_SENATE, session, roll_number)
            if path.exists():
                skipped += 1
                continue
            if fetched >= max_new:
                print(f"Reached --max-new {max_new}; remaining votes deferred to next run")
                return fetched, skipped
            url = senate_vote_source_url(congress, session, roll_number)
            try:
                vote = parse_senate_vote(_fetch(url), lis_to_bioguide)
                if vote["session"] != session or vote["roll_number"] != roll_number:
                    raise ValueError(
                        f"detail XML identifies itself as session {vote['session']} "
                        f"roll {vote['roll_number']}"
                    )
            except Exception as exc:  # noqa: BLE001 — per-vote isolation
                errors.record("senate_vote", f"{congress}-{session}-{roll_number}", exc, url=url)
                continue
            _common._write_json(path, vote)
            fetched += 1
            print(f"  + {vote['id']}: {vote['question']} — {vote['result']}")
    return fetched, skipped


def fetch_new_house_votes(
    congress: int,
    errors: ErrorCollector,
    max_new: int,
) -> tuple[int, int, int]:
    """Probe clerk.house.gov for new House roll calls and save them.

    The House publishes no vote menu, but roll numbers are dense and
    sequential per session, so each session is walked from roll 1: rolls
    already on disk are skipped for free, missing ones are fetched, and the
    first 404 is the end of the published roll calls. Candidate-ballot votes
    (Speaker elections) cannot be expressed as yea/nay positions; they are
    skipped — and re-probed every run, a couple of cheap fetches per Congress
    — rather than recorded, keeping file existence as the only resume cursor.

    Returns (fetched, skipped, unsupported).
    """
    fetched = 0
    skipped = 0
    unsupported = 0
    for session in HOUSE_SESSIONS:
        consecutive_errors = 0
        roll_number = 0
        while True:
            roll_number += 1
            path = vote_path(congress, CHAMBER_HOUSE, session, roll_number)
            if path.exists():
                skipped += 1
                continue
            if fetched >= max_new:
                print(f"Reached --max-new {max_new}; remaining votes deferred to next run")
                return fetched, skipped, unsupported
            url = house_vote_source_url(congress, session, roll_number)
            vote_key = f"{congress}-{session}-{roll_number}"
            try:
                text = _fetch(url)
            except requests.HTTPError as exc:
                if exc.response is not None and exc.response.status_code == 404:
                    break  # end of the published roll calls for this session
                errors.record("house_vote", vote_key, exc, url=url)
                break  # outage vs. end is indistinguishable; retry next run
            except requests.RequestException as exc:
                errors.record("house_vote", vote_key, exc, url=url)
                break
            try:
                vote = parse_house_vote(text)
            except UnsupportedVoteError as exc:
                unsupported += 1
                consecutive_errors = 0
                print(f"  - house-{vote_key} skipped: {exc}")
                continue
            except Exception as exc:  # noqa: BLE001 — per-vote isolation
                errors.record("house_vote", vote_key, exc, url=url)
                consecutive_errors += 1
                if consecutive_errors >= MAX_CONSECUTIVE_HOUSE_ERRORS:
                    print(
                        f"  ! {consecutive_errors} consecutive House parse failures; "
                        f"stopping session {session} probe"
                    )
                    break
                continue
            if (
                vote["congress"] != congress
                or vote["session"] != session
                or vote["roll_number"] != roll_number
            ):
                errors.record(
                    "house_vote",
                    vote_key,
                    ValueError(f"detail XML identifies itself as {vote['id']}"),
                    url=url,
                )
                break  # the year->session mapping is off; nothing later can be right
            consecutive_errors = 0
            _common._write_json(path, vote)
            fetched += 1
            print(f"  + {vote['id']}: {vote['question']} — {vote['result']}")
    return fetched, skipped, unsupported


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--congress", type=int, default=None,
        help="Congress number (default: the current Congress).",
    )
    parser.add_argument(
        "--max-new", type=int, default=1000,
        help="Cap on new votes fetched per run, shared across both chambers "
             "(default 1000). A full-Congress first backfill is ~1500+ detail "
             "fetches; steady-state runs fetch a handful. The cap bounds a "
             "runaway loop, and a capped run is harmless — the next run picks "
             "up where it left off.",
    )
    args = parser.parse_args(argv)
    congress = args.congress if args.congress is not None else current_congress()

    try:
        menus = fetch_senate_menus(congress)
        lis_to_bioguide = build_lis_to_bioguide()
    except Exception as exc:  # noqa: BLE001 — top-level reporter
        print(f"ERROR: {exc}", file=sys.stderr)
        return 1

    sessions = [m["session"] for m in menus]
    total_listed = sum(len(m["vote_numbers"]) for m in menus)
    print(
        f"Senate menu for Congress {congress}: {total_listed} roll calls "
        f"across session(s) {sessions}"
    )

    errors = ErrorCollector()
    senate_fetched, senate_skipped = fetch_new_senate_votes(
        congress, menus, lis_to_bioguide, errors, args.max_new
    )

    house_budget = args.max_new - senate_fetched
    if house_budget > 0:
        print(f"House probe for Congress {congress} (clerk.house.gov EVS):")
        house_fetched, house_skipped, unsupported = fetch_new_house_votes(
            congress, errors, house_budget
        )
    else:
        print("House probe deferred to next run: --max-new budget exhausted")
        house_fetched = house_skipped = unsupported = 0

    index = rebuild_votes_index(congress)
    errors.print_summary(label="fetch_votes")
    print(
        f"OK: {senate_fetched + house_fetched} new "
        f"({senate_fetched} Senate, {house_fetched} House), "
        f"{senate_skipped + house_skipped} already on disk, "
        f"{unsupported} unsupported, {len(errors)} failed; "
        f"index now {index['vote_count']} votes"
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
